components/inputs/input-number.js

```diff
@@ -99,42 +92,84 @@ class InputNumber extends SkeletonMixin(FormElementMixin(LocalizeCoreElement(Lit
 		this.maxExclusive = false;
 		this.minExclusive = false;
 		this.required = false;
-		this.trailingZeroes = false;
 
 		this._formattedValue = '';
 		this._inputId = getUniqueId();
-		this._numDecimalDigits = 0;
+		this._trailingZeroes = false;
+		this._valueTrailingZeroes = '';
+	}
+
+	get maxFractionDigits() {
+		return this._maxFractionDigits ?? Math.max(this.minFractionDigits, 3);
+	}
+	set maxFractionDigits(val) {
+		if (isNaN(val) || val < 0 || val > 20 || val < this.minFractionDigits) {
+			throw new RangeError('maxFractionDigits must be between 0 and 20 and >= minFractionDigits');
+		}
+		this._maxFractionDigits = val;
+		this._updateFormattedValue();
 	}
 
-	get value() { return this._value; }
+	get minFractionDigits() {
+		return this._minFractionDigits ?? 0;
+	}
+	set minFractionDigits(val) {
+		if (isNaN(val) || val < 0 || val > 20 || val < this.minFractionDigits) {
+			throw new RangeError('minFractionDigits must be between 0 and 20 and <= maxFractionDigits');
+		}
+		this._minFractionDigits = val;
+		this._updateFormattedValue();
+	}
+
+	get trailingZeroes() { return this._trailingZeroes; }
+	set trailingZeroes(val) {
+		this._trailingZeroes = val;
+		this._updateFormattedValue();
+	}
+
+	get value() {
+		if (this._value === undefined) return undefined;
+		const precision = Math.pow(10, this.maxFractionDigits);
+		return Math.round((this._value + Number.EPSILON) * precision) / precision;
+	}
 	set value(val) {
 		const oldValue = this.value;
-		try {
-			if (val === null) {
-				this._formattedValue = '';
-				this._value = undefined;
-			} else {
-				this._formattedValue = formatValue(
-					val,
-					this.minFractionDigits,
-					this.maxFractionDigits,
-					this.trailingZeroes,
-					this._numDecimalDigits
-				);
-				this._value = parseNumber(this._formattedValue);
-			}
-		} catch (err) {
-			this._formattedValue = '';
-			this._value = undefined;
+		if (val === null || isNaN(val)) {
+			val = undefined;
 		}
-		this._numDecimalDigits = 0;
+		this._value = val;
+		this._updateFormattedValue();
 		this.requestUpdate('value', oldValue);
 	}
 
+	get valueTrailingZeroes() {
+		if (this._valueTrailingZeroes === '') {
+			return '';
+		}
+		const numDecimals = Math.min(
+			countDecimalDigits(this._valueTrailingZeroes, false),
+			this.maxFractionDigits
+		);
+		let valueTrailingZeroes = this.value.toString();
+		const decimalDiff = (numDecimals - countDecimalDigits(valueTrailingZeroes, false));
+		if (decimalDiff > 0) {
+			if (decimalDiff === numDecimals) {
+				valueTrailingZeroes += '.';
+			}
+			valueTrailingZeroes = valueTrailingZeroes.padEnd(valueTrailingZeroes.length + decimalDiff, '0');
+		}
```

Comment:
The logic for the `valueTrailingZeroes` is essentially:
- Count how many decimals were in the thing the user typed in, but then don't let it be more than `maxFractionDigits`
- Count how many decimals are in `this.value` (the `Number`), which is already rounded properly
- If there are more decimals in `valueTrailingZeroes`, the user must have typed in a bunch of zeroes, so add that many to the end (and a `.` decimal symbol if needed)